Bluetooth: Audio: Add support for using bidirectional CIS for Audio Streams

Thalley opened the pull request:
This PR adds support for using a bidirectional CIS for a IN and OUT audio stream. The stack will couple streams together automatically, and will always attempt to use the least amount of CIS. 

fixes https://github.com/zephyrproject-rtos/zephyr/issues/41193

Thalley committed:
Bluetooth: Audio: Move ISO struct from audio_endpoint

Moves the bt_iso_chan and bt_iso_chan_qos structs
out from the bt_audio_ep struct, and instead use a
reference.

This is due to the fact that an endpoint can ever
only be unidirectional, and a CIS can be bidrectional.

The support for using a single bididrectional CIS
for two audio streams/endpoints is not implemented
in this commit.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Add sink/source ep ref to bt_audio_iso

Add supporting for referercing either a sink endpoint,
a source endpoint or both in the bt_audio_iso struct.

This way a single bt_audio_iso can have references to up
to 2 endpoints.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Use bidirectional CIS for unicast audio

The unicast client and server will attempt to use a
bidirectional CIS instead of two unidirectional CIS
whenever possible.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
Bluetooth: Audio: Make bt_audio_codec_qos_to_iso_qos return void

Change the return type of bt_audio_codec_qos_to_iso_qos to
void as it cannot fail, and this will make it easier to use.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
samples: Bluetooth: Update unicast audio client to multi-cis

Add support for setting up, configuring, and using
multiple CIS in the sample.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
samples: Bluetooth: unicast_audio_server multi-cis support

Add support for multiple CIS.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley committed:
samples: Bluetooth: Add tx support for unicast audio server

Add support to send mock data on any source endpoint/streams
configured by the client.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>

Thalley commented:
Depends on https://github.com/zephyrproject-rtos/zephyr/pull/43433

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
Rebased to solve merge conflicts

Thalley commented:
To fix https://github.com/zephyrproject-rtos/zephyr/issues/45374 the allocation of ISO channels (from bt_unicast_client_new_audio_iso), needs to be done when creating the group or when added streams to the group, otherwise we can't create the CIG without all streams being configured first.

MariuszSkamra commented on samples/bluetooth/unicast_audio_client/src/main.c:
```diff
@@ -170,8 +181,8 @@ static void init_lc3(void)
 	octets_per_frame = bt_codec_cfg_get_octets_per_frame(&codec_configuration.codec);
-	frames_per_sdu = bt_codec_cfg_get_frame_blocks_per_sdu(audio_stream.codec, true);
-	octets_per_frame = bt_codec_cfg_get_octets_per_frame(audio_stream.codec);
+	frames_per_sdu = bt_codec_cfg_get_frame_blocks_per_sdu(streams[0].codec, true);
+	octets_per_frame = bt_codec_cfg_get_octets_per_frame(streams[0].codec);
```
Why those cannot be obtained from `&codec_configuration.codec` the same way as parameters above?

Thalley replied:
They should be - Updated :)

MariuszSkamra commented on samples/bluetooth/unicast_audio_client/src/main.c:
```diff
@@ -245,17 +256,25 @@ static void audio_timer_timeout(struct k_work *work)
+	 * data going to the server)
+	 */
+	for (size_t i = 0; i < configured_sink_stream_count; i++) {
+		buf = net_buf_alloc(&tx_pool, K_FOREVER);
+		net_buf_reserve(buf, BT_ISO_CHAN_SEND_RESERVE);
```
Looks good, but probably you could reuse the buffer, by just bumping its reference

Thalley replied:
Oh yeah, that's actually a good idea since we are sending the same SDU on multiple CIS. Thanks!

MariuszSkamra commented on samples/bluetooth/unicast_audio_server/src/main.c:
```diff
@@ -131,6 +135,66 @@ static void print_qos(struct bt_codec_qos *qos)
+ *
+ * @param work Pointer to the work structure
+ */
+static void audio_timer_timeout(struct k_work *work)
```
Do you plan to add LC3 encoder as well?

Thalley replied:
That is probably a good idea, but the change in this sample was mostly to be able to test the stack change. The LC3 support should be added in a separate PR: https://github.com/zephyrproject-rtos/zephyr/issues/45595

MariuszSkamra commented on subsys/bluetooth/audio/ascs.c:
```diff
@@ -51,6 +51,10 @@ struct bt_ascs {
+	/* A single iso_channel may be used for 1 or 2 ases.
+	 * Controlled by the client.
+	 */
+	struct bt_audio_iso isos[ASE_COUNT];
```
Why the ISO is not a member of `bt_ascs_ase`?

Thalley replied:
That is also an option. The reason why it isn't, is that two ASE may use the same CIS and thus the same `struct bt_audio_iso`. In other words, there isn't a 1:1 between ASE and `struct bt_audio_iso`, and `struct bt_audio_iso isos[ASE_COUNT];` may also in the future to be less than `ASE_COUNT` (e.g. if we want to support fewer started streams than ASEs).

MariuszSkamra replied:
Indeed. One ASE can be assigned to only one CIS, but one CIS can be assigned to multiple ASE's. Thus it's intuitive that each ASE keeps assigned CIS

Thalley replied:
Not sure I follow your comment. 

If we move the ISO into the `bt_ascs_ase` then we cannot have 2 ASEs with the same ISO.

MariuszSkamra replied:
Now I see it. I thought it's pointer rather than value. Sorry :)

MariuszSkamra reviewed (changes requested):
Much appreciate your PR, I will give it a try. Some comments added

Thalley commented:
Rebased to solve merge conflicts

Thalley commented on samples/bluetooth/unicast_audio_client/src/main.c:
```diff
@@ -19,14 +19,20 @@ static void start_scan(void);
 static struct bt_audio_ep *sinks[CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SNK_COUNT];
-NET_BUF_POOL_FIXED_DEFINE(tx_pool, 1,
+static struct bt_audio_ep *sources[CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SRC_COUNT];
+NET_BUF_POOL_FIXED_DEFINE(tx_pool, CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SNK_COUNT,
```
`CONFIG_BT_AUDIO_UNICAST_CLIENT_ASE_SNK_COUNT` is correct, as `SINK` refers to remote sinks. 

This is actually something that has been bothering me a little:

Should "sink" refer to a local sink (i.e. remote source ASE), or a remote sink ASE?

Thalley commented:
Rebased to solve merge conflicts

Thalley committed:
Bluetooth: Audio: Fix issues with unicast audio group lifetime

The unicast audio group can not have streams removed after the
CIG has been created as part of the QoS set procedure.

The bt_audio_unicast_group_remove_streams function may
leave the unicast group without streams, and thus the
check for "free" group was not correct, and has been replaced
with a simple boolean value.

Signed-off-by: Emil Gydesen <emil.gydesen@nordicsemi.no>
